XenModel/Network/Heartbeat.cs

```diff
@@ -148,6 +148,23 @@ private void DoHeartbeat()
                 HandleConnectionLoss();
                 return;
             }
+            catch (WebException exn)
+            {
+                log.Error(exn);
+                if (((HttpWebResponse)exn.Response).StatusCode == HttpStatusCode.ProxyAuthenticationRequired) // work-around for CA-214653
+                {
+                    log.DebugFormat("Heartbeat for {0} has failed due to {1} credentials; closing the main connection",
+                        session == null ? "null" : session.Url,
+                        session.proxy.Proxy.Credentials == null ? "missing" : "incorrect");
+                    connection.Interrupt();
+                    DropSession();
+                }
+                else
+                {
+                    HandleConnectionLoss();
+                }
+                return;

```

Comment:
The returns at the end of each catch case are not necessary, it will return anyway since there is nothing to do after the try-catch and before the end of the method.